java/HealthVault Mobile/HVSample/src/main/java/com/microsoft/hsg/android/hvsample/WeightActivity.java

```diff
@@ -0,0 +1,163 @@
+    
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.weight);
+        mService = HealthVaultApp.getInstance();
+        mHVClient = new HealthVaultClient();
+        
+        Button weightsBtn = (Button) findViewById(R.id.addWeight);
+        final EditText editText = (EditText) findViewById(R.id.weightInput);
+		
+        weightsBtn.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View view) {
+            	if (mService.isAppConnected()) {
+                    try {
+                        putWeight(editText.getText().toString());
+                        editText.setText("");
+                    }
+                    catch (Exception e) {
+                        Toast.makeText(WeightActivity.this, "Please enter a weight!", Toast.LENGTH_SHORT).show();
+                    }
+            	}
+            }
+        });
+
+        setTitle("Weight Sample");
+    }
+    
+	@Override
+	protected void onStart() {
+		super.onStart();
+        mHVClient.start();
+	}
+    
+    @Override
+    protected void onResume()
+    {
+        super.onResume();
+        mCurrentRecord = HealthVaultApp.getInstance().getCurrentRecord();
+        getWeights();
+    }
+    
+    @Override
+	protected void onStop() {
+        mHVClient.stop();
+		super.onStop();
+	}
+
+    @SuppressWarnings("unchecked")
+    private void getWeights()
+    {
+        mHVClient.asyncRequest(
+                mCurrentRecord.getThingsAsync(ThingRequestGroup2.thingTypeQuery(Weight.ThingType)),
+    			new WeightCallback(WeightCallback.RenderWeights));
+    }
+    
+    private void putWeight(String value) {
+        final Thing2 thing = new Thing2();
+		thing.setData(new Weight(Double.parseDouble(value)));
+        // thing.setData(new Weight(-10));
+        mHVClient.asyncRequest(
+                mCurrentRecord.putThingAsync(thing),
+				new WeightCallback(WeightCallback.PutWeights));
+    }
+    
+    private void renderWeights(List<Thing2> things) {
+    	List<String> weights = new ArrayList<String>();
+        int count = 0;
+
+        for(Thing2 thing : things) {
+        	Weight w = (Weight)thing.getData();
+            if (count < 1)
+            {
+                TextView lastWeight = (TextView) findViewById(R.id.lasWeight);
+                lastWeight.setText(String.valueOf(w.getValue().getKg()));
+                count++;
+            }
+        	weights.add(String.valueOf(w.getWhen().getDate().getM() + "/" +
+                            w.getWhen().getDate().getD() + "/" +
+                            w.getWhen().getDate().getY()) + "               " +
+                            "                    " +
+                    String.valueOf(w.getValue().getKg()
+                    ));
+        }
+        ListView lv = (ListView)findViewById(R.id.weightList);
+        lv.setAdapter(new ArrayAdapter<String>(
```

Comment:
Each time we resume we call get weights, each time we getWeights we call to render here. We can then end up calling this code twice for the same instance of the listview. This is problematic because calling setAdapter when an adapter is already set can cause a crash. Instead you should find the view and create the adapter somewhere earlier (onCreate probably) and then simply update the adapter contents here. The adapter is bound to the listview so all you need to do is update the adapter contents and then call notifyDatasetChanged() on the adapter and the listview will will handle the change.